src/main/io/cms.c

```diff
@@ -0,0 +1,1342 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Cleanflight is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Cleanflight.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ Created by Marcin Baliniak
+ OSD-CMS separation by jflyper
+ */
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <string.h>
+#include <ctype.h>
+
+#include "platform.h"
+
+#include "build/version.h"
+
+#ifdef CMS
+
+#include "drivers/system.h"
+
+#include "common/typeconversion.h"
+
+#include "io/cms.h"
+#include "io/cms_types.h"
+#ifdef CANVAS
+#include "io/canvas.h"
+#endif
+
+#include "io/flashfs.h"
+#include "io/osd.h"
+#include "io/display.h"
+
+#include "fc/config.h"
+#include "fc/rc_controls.h"
+#include "fc/runtime_config.h"
+
+#include "flight/pid.h"
+
+#include "config/config_profile.h"
+#include "config/config_master.h"
+#include "config/feature.h"
+
+#include "io/cms.h"
+
+#include "build/debug.h"
+
+// Device management
+
+#define CMS_MAX_DEVICE 4
+
+cmsDeviceInitFuncPtr cmsDeviceInitFunc[CMS_MAX_DEVICE];
+int cmsDeviceCount;
+int cmsCurrentDevice = -1;
+int cmsLastDevice = -1;
+
+bool cmsDeviceRegister(cmsDeviceInitFuncPtr func)
+{
+    if (cmsDeviceCount == CMS_MAX_DEVICE)
+        return false;
+
+    cmsDeviceInitFunc[cmsDeviceCount++] = func;
+
+    return true;
+}
+
+cmsDeviceInitFuncPtr cmsDeviceSelectCurrent(void)
+{
+    if (cmsDeviceCount == 0)
+        return NULL;
+
+    if (cmsCurrentDevice < 0)
+        cmsCurrentDevice = 0;
+
+    return cmsDeviceInitFunc[cmsCurrentDevice];
+}
+
+cmsDeviceInitFuncPtr cmsDeviceSelectNext(void)
+{
+    if (cmsDeviceCount == 0)
+        return NULL;
+
+    cmsCurrentDevice = (cmsCurrentDevice + 1) % cmsDeviceCount; // -1 Okay
+
+    return cmsDeviceInitFunc[cmsCurrentDevice];
+}
+
+#define CMS_UPDATE_INTERVAL 50 // msec
+
+// XXX Why is this here? Something wrong?
+// XXX Something like Drawing Context that holds all state variables would be the way...
+int8_t lastCursorPos;
+
+void cmsScreenClear(displayPort_t *instance)
+{
+    instance->vTable->clear();
+    instance->cleared = true;
+    lastCursorPos = -1; // XXX Here
+}
+
+void cmsScreenBegin(displayPort_t *instance)
+{
+    instance->vTable->begin();
+    instance->vTable->clear();
+}
+
+void cmsScreenEnd(displayPort_t *instance)
+{
+    instance->vTable->end();
+}
+
+int cmsScreenWrite(displayPort_t *instance, uint8_t x, uint8_t y, char *s)
+{
+    return instance->vTable->write(x, y, s);
+}
+
+void cmsScreenHeartBeat(displayPort_t *instance)
+{
+    if (instance->vTable->heartbeat)
+        instance->vTable->heartbeat();
```

Comment:
Rather than have `if (instance->vTable->heartbeat)` just call `instance->vTable->heartbeat();` and ensure vTable is set up with an empty `heartbeat` function - the overhead is about the same and the code is much neater. Likewise with `resync` and `txroom`